pkg/clustermesh/common/clustermesh.go

```diff
@@ -108,13 +118,17 @@ func (cm *clusterMesh) Start(cell.HookContext) error {
 // Close stops watching for remote cluster configuration files to appear and
 // will close all connections to remote clusters
 func (cm *clusterMesh) Stop(cell.HookContext) error {
-	cm.mutex.Lock()
-	defer cm.mutex.Unlock()
-
 	if cm.configWatcher != nil {
 		cm.configWatcher.close()
 	}
 
+	// Wait until all in-progress removal processes have completed, if any.
+	// We must not hold the mutex at this point, as needed by the go routines.
+	cm.wg.Wait()
```

Comment:
I think we need to distinguish different cases here:

* Plain agent or kvstoremesh shutdown, with no clustermesh configurations removals (i.e., the standard case): clustermesh just stops the watchers and closes the etcd client. This was, and continues to be, performed synchronously by the stop hook, and is expected to be instantaneous (unless the agent is blocked processing events already received).
* agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
* kvstoremesh shutdown while the removal of one or more clusters is being processed. In that case, [the context of the job should be closed](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-delete-cluster/pkg/clustermesh/kvstoremesh/remote_cluster.go#L183), hence causing the removal logic to return immediately, and unblock the wait group. Now that I'm writing this, I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

Reply:
> I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

Fixed with the latest push: https://github.com/cilium/cilium/compare/81721f9e913c55d75137a74c4bb0ec45d5815fc2..a5ad5df50a339422506088edab0b67c68336715a